### ISSUE TITLE
Hardware transcoder failover/priority supported?

### ISSUE THREAD

**Issue by Ornias1993** [jellyfin/jellyfin#4389]
Before I write a feature request or another kinda of report...
I was looking into failover/priority of hardware transcoders, before offloading to the CPU

**Example**
One could have both a Nvidia GPU and a intel quicksync compatible CPU.
In those cases it makes sense to first try to encode on the Nvidia GPU and if those slots are full, transcode on using quicksync. In a lot of systems this would almost double(!) the amount of available transcodes.

I tried to look into the code if and how such a feature is currently implemented, does anyone have any input on this?

**Comment by dkanada** [comment]
We don't even failover to software decoding when the selected hardware option breaks. Any kind of implementation would be welcome, especially if it fixed the above issue.

**Comment by Ornias1993** [comment]
@dkanada Thanks, I edited the Issue accordingly.

**Comment by nyanmisaka** [comment]
It's possible to add such feature. User can custom the priority of encoders.

On customer grade Nvidia cards, NVENC concurrent sessions are limited to only three without using the additional patch.

We need a fallback feature to do `NVENC(3 sessions/failed) -> QuickSync(failed) -> libx264(software)`

**Comment by RazeLighter777** [comment]
It would be interesting if you could implement a customizable priority list of transcoding methods. When ffmpeg returns an error, it simply uses the next highest transcoding method on the list. This would be more flexible and also prevent errors from unsupported codecs in hardware transcoding.

**Comment by dkanada** [comment]
@nyanmisaka the new feature @joshuaboniface added sounds like the perfect method to track NVENC streams.

**Comment by Ornias1993** [comment]
@nyanmisaka @RazeLighter777 
If it's possible to even make it configurable, that would be even better ofc.
Even better would be if we could also add a max-transcode count to every transcoder in the list. For use with patched NVENC drivers for example

But indeed, the NVENC limits was one of my prime reasons to throw this up. :)

**Comment by nyanmisaka** [comment]
> @nyanmisaka the new feature @joshuaboniface added sounds like the perfect method to track NVENC streams.

That only limit the maximum number of active sessions per account, exceeded users will not allowed to login.
We need to limit the transcoding sessions.

**Comment by Ornias1993** [comment]
> > @nyanmisaka the new feature @joshuaboniface added sounds like the perfect method to track NVENC streams.
> 
> That only limit the maximum number of active sessions per account, exceeded users will not allowed to login.
> We need to limit the transcoding sessions.

Agreed, better we do it right the first time and add both features... because basically it would be the same codepath anyway.

**Comment by nyanmisaka** [comment]
Current steps of transcoding: if ffmpeg failed to run, jellyfin will retry 3 times and throw errors to the client.

HLS requests -> DynamicHlsController.cs -> GetDynamicSegment() -> GetSegmentResult() -> StartFfMpeg(ffmpeg_cli)

EncodingHelper.cs -> GetCommandLineArguments() -> ffmpeg_cli

**Comment by Ornias1993** [comment]
> Current steps of transcoding: if ffmpeg failed to run, jellyfin will retry 3 times and throw errors to the client.
> 
> HLS requests -> DynamicHlsController.cs -> GetDynamicSegment() -> GetSegmentResult() -> StartFfMpeg(ffmpeg_cli)
> 
> EncodingHelper.cs -> GetCommandLineArguments() -> ffmpeg_cli

Cool, so initially we could just cycle through that for every transcoding option available...

**Comment by nyanmisaka** [comment]
The problem is that each try will create a whole new StreamState(just like session), jellyfin doesn't know if the last try was failed.

**Comment by Ornias1993** [comment]
@nyanmisaka I don't think we even need to know if the last try failed, only if the 3 try's failed.

If you can proces the error, you can instead of throwing errors to the client, set a different transcoder.
You only need to add keeping track of the actual transcoder.

**Comment by nyanmisaka** [comment]
Need to correct a little. I found that the number of retries depends on the HLS player.
It will not retry on edge chromium, but will keep retrying on ios14 safari.

**Comment by Ornias1993** [comment]
@nyanmisaka Okey that means it would indeed be a little more work implementing this....

**Comment by RazeLighter777** [comment]
Note failovers should definitely be logged or made known to the admin somehow, as they might lead some users to think their hardware transcoding is working when it really is not just because it fails quietly.

**Comment by nyanmisaka** [comment]
This is the reason why I spent a few months getting the various ffmpeg commands for HWA correct.

Client should retry and tell the server the retry count. Then server can use it and prioritisation to execute the next encoder option.

**Comment by nyanmisaka** [comment]
> Note failovers should definitely be logged or made known to the admin somehow, as they might lead some users to think their hardware transcoding is working when it really is not just because it fails quietly.

Or can be a toast to let user know.

**Comment by nyanmisaka** [comment]
Related options in hls.js.
https://github.com/video-dev/hls.js/blob/master/docs/API.md#fragloadingmaxretry--manifestloadingmaxretry--levelloadingmaxretry

**Comment by dkanada** [comment]
Toasts are indeed the best option for the web client.

**Comment by PrplHaz4** [comment]
From what I understand, this might also solve this issue with remote transcode jobs not having/needing the same HWA settings: https://github.com/joshuaboniface/rffmpeg/issues/3

FFMPEG would be restarted with the new parameters, and if supported on the remote transcode host it would succeed, else fallbck until it gets parameters that work...

**Comment by Ornias1993** [comment]
@PrplHaz4 I think that issue is a bit more complex, but yes, that could/should be easier to implement once this feature is implemented. :)

**Comment by Ornias1993** [comment]
As far as I'm aware, Jellyfin still does not have this feature-parity with other solutions at this date.
I've not forgoten the issue and am still waiting for this before I can take moving to jellyfin into consideration more serieusly.

**Comment by themcv** [comment]
Still interested in this. I am attempting to cluster Jellyfin, and the other nodes would not have an NVENC decoder and would need to fallback to QuickSync.

**Comment by Mavyre** [comment]
Could it be quickfixed with a wrapper script while this is being implemented natively into Jellyfin? In #3929, people are using a wrapper script to make it work. That would require to domodify on the fly some arguments given by Jellyfin to replace it with the software transcode "version".

Following something like `ffmpeg -hwaccel vdpau -hwaccel_device /dev/dri/renderD128 -i input.mp4 output.mp4 || ffmpeg -i input.mp4 output.mp4`, we could make it fallback automatically on failure. But we need a bit more information/digging about how Jellyfin uses ffmpeg to modify codec and disable hardware acceleration.

That would allow to still use rffmpeg too: if the remote hwaccel fails, it fails over to remote software transcoding. And if the remote is offline, to the local software transcoding.

**Comment by Mavyre** [comment]
As an alternative (and first iteration), could we just modify [VideosController.cs](../blob/master/Controllers/VideosController.cs) to allow a basic software failover?

```C#

try
{
    var encodingOptions = _serverConfigurationManager.GetEncodingOptions();
    var ffmpegCommandLineArguments = _encodingHelper.GetProgressiveVideoFullCommandLine(state, encodingOptions, outputPath, "superfast");
    return await FileStreamResponseHelpers.GetTranscodedFile(
        state,
        isHeadRequest,
        HttpContext,
        _transcodingJobHelper,
        ffmpegCommandLineArguments,
        _transcodingJobType,
        cancellationTokenSource).ConfigureAwait(false);
}
catch(Exception ex)
{
    encodingOptions.EnableHardwareEncoding = false;
    var ffmpegCommandLineArguments = _encodingHelper.GetProgressiveVideoFullCommandLine(state, encodingOptions, outputPath, "superfast");
    return await FileStreamResponseHelpers.GetTranscodedFile(
        state,
        isHeadRequest,
        HttpContext,
        _transcodingJobHelper,
        ffmpegCommandLineArguments,
        _transcodingJobType,
        cancellationTokenSource).ConfigureAwait(false);
}
```

In the future, we could even change `encodingOptions.HardwareAccelerationType` to QSV (that would allow to do what @themcv wants to achieve) to failover instead of disabling completely HW Acceleration.

But that would require to modify user interface, and I don't have enough knowledge both in C# and Jellyfin code yet to make a pull request. If anyone can help, that would be much appreciated!

**Comment by Mavyre** [comment]
Shouldn't go to stale if it's in a triage board, should it?

**Comment by nyanmisaka** [comment]
I'll be spending some time on it this year. This requires many changes, including ffmpeg, server, and web.

**Comment by Ornias1993** [comment]
> I'll be spending some time on it this year. This requires many changes, including ffmpeg, server, and web.

Yeah it's significant AF.
If you need any assistance or testing, I could as the TrueCharts staff to tryout some testing code before release!

**Comment by nyanmisaka** [comment]
> > I'll be spending some time on it this year. This requires many changes, including ffmpeg, server, and web.
> 
> Yeah it's significant AF. If you need any assistance or testing, I could as the TrueCharts staff to tryout some testing code before release!

I'll let you know when we have some new stuff.

**Comment by akmail** [comment]
what about automated recognition of what the server is capable of? maybe with a button in the settings area to re-scan the capabilities? would require some smaller sample files probably to be included in the server. Ideally it would be nice to not let users decide what his/her server is capable of - jellyfin should find it out automatically. the idea above is nice though as well.

**Comment by geekpete** [comment]
Even without changing ffmpeg, if the server can just detect one of the known ffmpeg transcode error numbers occurred (or even just any non-success exit code occurred) and immediately retry in software only "None" transcode mode, this would likely cover many scenarios and succeed more often than not. Even if this isn't a default or starts as an experimental option that can be enabled.
Having only software ("None") or a specific hardware encoder doesn't deal well with a mixed format/codec library.

Is it worth a separate issue to allow per video file/file type/media library specific override as well? For example if you play a file and it fails, manually set transcode preference just for that file (eg back to "None" for hardware acceleration transcode) to allow it to play from then on.

**Comment by nyanmisaka** [comment]
Here are some updates https://github.com/jellyfin/jellyfin-ffmpeg/pull/733

Modifying ffprobe/ffmpeg is unavoidable, given the increasing complexity of our transcoding pipeline and the need of 10bit/HDR encoding. Without these changes, we would lack the necessary information to determine the number of GPUs on the system, the specific HWA methods they support, and the supported codecs along with their attributes (such as min/max_width/height, profiles, and pixel formats).

The modified ffprobe now supports outputting the required GPU capabilities as a JSON string. The server simply parses this output, eliminating the need to manually select codecs via checkboxes and enabling intelligent prioritization based on GPU capabilities and utilizations. Ultimately, the server can provide a GPU list, which users can override according to their preferences.

If anyone is interested in testing this ffprobe, feel free to check it out and give it a try. You are welcome to test this in complex multi-GPU environments, please let me know if the output looks abnormal or if it crashes. We will begin server-side changes once it is sufficiently reliable.
